File: python/maze.py
# Source of truth: shared/game_constants.json

# ═══════════════════════════════════════════════════════
# EDUCATIONAL NOTE: Rendering/Drawing
# ═══════════════════════════════════════════════════════
# The Maze class is a PURE DATA MODEL with zero rendering imports.
# It stores wall/pellet state only; the rendering layer (game_panel.py)
# reads this data via is_wall() / has_pellet() and draws tiles itself.
#
# Compare with:
#   Java (Swing) → In many Java tutorials, maze rendering is embedded
#   directly inside a JPanel's paintComponent() alongside the data.
#   Here Python encourages a cleaner separation because there is no
#   required class hierarchy (no "extends JPanel").
#
# KEY CONCEPT: Separating data from drawing makes classes easier to
# test and reuse -- a core benefit of Python's flexible design.
# ═══════════════════════════════════════════════════════


class Maze:
    """Tile-based level data: walls and pellets on a 40x30 grid.

    Pure data model with ZERO rendering imports. The rendering layer
    reads wall/pellet state via is_wall()/has_pellet() and draws
    the maze itself.
    """

    _MAZE_WIDTH = 40
    _MAZE_HEIGHT = 30

    def __init__(self):
        self._walls: list[list[bool]] = []
        self._pellets: list[list[bool]] = []
        self._initialize_maze()

    # ------------------------------------------------------------------
    # Maze construction
    # ------------------------------------------------------------------

    def _initialize_maze(self) -> None:
        # ═══════════════════════════════════════════════════════
        # EDUCATIONAL NOTE: Object Movement
        # ═══════════════════════════════════════════════════════
        # The maze grid is the foundation for ALL movement. Every entity
        # (Pacman, Ghosts) moves tile-by-tile and checks this grid before
        # stepping. The 2D list is built with a list comprehension --
        # one line creates 40 columns, each containing 30 booleans.
        #
        # Compare with:
        #   Java (Swing) → boolean[][] walls = new boolean[40][30];
        #   Java requires explicit array type declarations and sizes.
        #   Python's list comprehension is more concise and readable.
        #
        # KEY CONCEPT: List comprehensions like [expr for _ in range(n)]
        # replace verbose Java loops for initializing 2D arrays.
        # ═══════════════════════════════════════════════════════

        # Allocate grids – indexed as [x][y] to match Java reference
        self._walls = [
            [False] * self._MAZE_HEIGHT for _ in range(self._MAZE_WIDTH)
        ]
        self._pellets = [
            [False] * self._MAZE_HEIGHT for _ in range(self._MAZE_WIDTH)
        ]

        # Border walls – top / bottom
        for x in range(self._MAZE_WIDTH):
            self._walls[x][0] = True
            self._walls[x][self._MAZE_HEIGHT - 1] = True

        # Border walls – left / right
        for y in range(self._MAZE_HEIGHT):
            self._walls[0][y] = True
            self._walls[self._MAZE_WIDTH - 1][y] = True

        # Internal walls: horizontal rows at y=10 and y=20, x in [5, 15)
        for x in range(5, 15):
            self._walls[x][10] = True
            self._walls[x][20] = True

        # Internal wall: vertical column at x=20, y in [5, 25)
        for y in range(5, 25):
            self._walls[20][y] = True

        # ═══════════════════════════════════════════════════════
        # EDUCATIONAL NOTE: Collision Detection
        # ═══════════════════════════════════════════════════════
        # Pellets are placed on every open (non-wall) interior tile.
        # Later, collision detection in game_logic.py checks whether
        # Pacman's grid position matches a pellet tile. This simple
        # "same tile = collision" approach works because movement is
        # tile-based (discrete), not pixel-based (continuous).
        #
        # Compare with:
        #   Java (Swing) → The same grid-based check, but Java would
        #   typically use nested for-loops with int i/j counters.
        #   Python's range() + boolean expressions are more compact.
        #
        # KEY CONCEPT: Grid-based collision is the simplest form --
        # two objects collide when they occupy the same (x, y) cell.
        # ═══════════════════════════════════════════════════════

        # Pellets on every open interior tile except Pacman start (5, 5)
        for x in range(1, self._MAZE_WIDTH - 1):
            for y in range(1, self._MAZE_HEIGHT - 1):
                if not self._walls[x][y] and not (x == 5 and y == 5):
                    self._pellets[x][y] = True

    # ------------------------------------------------------------------
    # Public query helpers
    # ------------------------------------------------------------------

    # ═══════════════════════════════════════════════════════
    # EDUCATIONAL NOTE: Collision Detection
    # ═══════════════════════════════════════════════════════
    # is_wall() is called by EVERY entity before moving. It treats
    # out-of-bounds coordinates as walls (fail-safe boundary check).
    # This single method centralizes all wall collision logic.
    #
    # Compare with:
    #   Java (Swing) → Same logic, but in Java you would write:
    #   if (x < 0 || x >= WIDTH || y < 0 || y >= HEIGHT) return true;
    #   Python uses English-like 'or' instead of '||' and does not
    #   need explicit return type declarations (though type hints
    #   like -> bool are optional and used here for clarity).
    #
    # KEY CONCEPT: Centralizing boundary checks in one method
    # prevents off-by-one bugs across Pacman and Ghost movement.
    # ═══════════════════════════════════════════════════════

    def is_wall(self, x: int, y: int) -> bool:
        """Return True if (x, y) is a wall **or** out of bounds."""
        if x < 0 or x >= self._MAZE_WIDTH or y < 0 or y >= self._MAZE_HEIGHT:
            return True
        return self._walls[x][y]

    def has_pellet(self, x: int, y: int) -> bool:
        """Return True if there is an uneaten pellet at (x, y)."""
        if x < 0 or x >= self._MAZE_WIDTH or y < 0 or y >= self._MAZE_HEIGHT:
            return False
        return self._pellets[x][y]

    def remove_pellet(self, x: int, y: int) -> None:
        """Remove the pellet at (x, y) if within bounds."""
        if 0 <= x < self._MAZE_WIDTH and 0 <= y < self._MAZE_HEIGHT:
            self._pellets[x][y] = False

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    # ═══════════════════════════════════════════════════════
    # EDUCATIONAL NOTE: Rendering/Drawing
    # ═══════════════════════════════════════════════════════
    # Properties like width/height let the rendering layer
    # (game_panel.py) iterate over the grid without knowing
    # internal constants. Python's @property decorator turns
    # a method into a read-only attribute access (maze.width).
    #
    # Compare with:
    #   Java (Swing) → public int getWidth() { return MAZE_WIDTH; }
    #   Java uses explicit getter methods. Python's @property
    #   gives the same encapsulation with cleaner call syntax:
    #   maze.width instead of maze.getWidth().
    #
    # KEY CONCEPT: @property is Python's idiomatic replacement for
    # Java-style getter/setter methods.
    # ═══════════════════════════════════════════════════════

    @property
    def width(self) -> int:
        return self._MAZE_WIDTH

    @property
    def height(self) -> int:
        return self._MAZE_HEIGHT

File: website/frontend/public/game/js/maze.js
// Source of truth: shared/game_constants.json
// Pure logic module — ZERO rendering/GUI code.

// ═══════════════════════════════════════════════════════
// EDUCATIONAL NOTE: Rendering/Drawing
// ═══════════════════════════════════════════════════════
// In JavaScript, the Maze class stores wall and pellet data as 2D arrays
// of booleans. The rendering layer (game.js) reads these arrays and draws
// them onto an HTML5 <canvas> using the Canvas 2D API (fillRect, arc).
// The data and drawing are fully separated across files.
//
// Compare with:
//   Java → Uses a boolean[][] with the same separation; rendering is done
//          via Graphics2D.fillRect() in a Swing JPanel's paintComponent().
//   Python → Uses a list-of-lists of booleans; rendering happens through
//            pygame.draw.rect() and pygame.draw.circle() in a separate loop.
//
// KEY CONCEPT: JavaScript's approach mirrors MVC — the Maze is pure data
// (model), while the Canvas 2D context in game.js is the view. No DOM or
// Canvas references exist in this file.
// ═══════════════════════════════════════════════════════

const MAZE_WIDTH = 40;
const MAZE_HEIGHT = 30;

class Maze {
    constructor() {
        this._walls = [];
        this._pellets = [];
        this._initializeMaze();
    }

    // ═══════════════════════════════════════════════════════
    // EDUCATIONAL NOTE: Object Movement
    // ═══════════════════════════════════════════════════════
    // The maze's wall data serves as the boundary system for ALL object
    // movement. Both Pacman and Ghost call maze.isWall(x, y) before
    // moving. JavaScript uses simple boolean arrays indexed by grid
    // coordinates — no spatial data structures or physics engines needed.
    //
    // Compare with:
    //   Java → Identical boolean[][] approach; isWall() is called by
    //          Pacman.update() and Ghost.update() with the same logic.
    //   Python → Uses a list-of-lists; is_wall() performs the same
    //            bounds check and boolean lookup.
    //
    // KEY CONCEPT: Grid-based collision boundaries are language-agnostic.
    // The simplicity of a boolean grid makes wall checks O(1) regardless
    // of maze complexity — the same pattern works in JS, Java, and Python.
    // ═══════════════════════════════════════════════════════
    _initializeMaze() {
        // Create 2D arrays initialized to false
        for (let x = 0; x < MAZE_WIDTH; x++) {
            this._walls[x] = new Array(MAZE_HEIGHT).fill(false);
            this._pellets[x] = new Array(MAZE_HEIGHT).fill(false);
        }

        // Create border walls
        for (let x = 0; x < MAZE_WIDTH; x++) {
            this._walls[x][0] = true;
            this._walls[x][MAZE_HEIGHT - 1] = true;
        }
        for (let y = 0; y < MAZE_HEIGHT; y++) {
            this._walls[0][y] = true;
            this._walls[MAZE_WIDTH - 1][y] = true;
        }

        // Add internal walls — horizontal at rows 10 and 20, columns 5-14 inclusive
        for (let x = 5; x < 15; x++) {
            this._walls[x][10] = true;
            this._walls[x][20] = true;
        }

        // Add internal wall — vertical at column 20, rows 5-24 inclusive
        for (let y = 5; y < 25; y++) {
            this._walls[20][y] = true;
        }

        // Initialize pellets on all non-wall tiles except Pacman start (5,5)
        for (let x = 1; x < MAZE_WIDTH - 1; x++) {
            for (let y = 1; y < MAZE_HEIGHT - 1; y++) {
                if (!this._walls[x][y] && !(x === 5 && y === 5)) {
                    this._pellets[x][y] = true;
                }
            }
        }
    }

    /**
     * Returns true if the tile at (x, y) is a wall.
     * Out-of-bounds coordinates are treated as walls.
     */
    isWall(x, y) {
        if (x < 0 || x >= MAZE_WIDTH || y < 0 || y >= MAZE_HEIGHT) {
            return true;
        }
        return this._walls[x][y];
    }

    /**
     * Returns true if the tile at (x, y) has an uneaten pellet.
     * Out-of-bounds coordinates return false.
     */
    hasPellet(x, y) {
        if (x < 0 || x >= MAZE_WIDTH || y < 0 || y >= MAZE_HEIGHT) {
            return false;
        }
        return this._pellets[x][y];
    }

    /**
     * Removes the pellet at (x, y) if within bounds.
     */
    removePellet(x, y) {
        if (x >= 0 && x < MAZE_WIDTH && y >= 0 && y < MAZE_HEIGHT) {
            this._pellets[x][y] = false;
        }
    }

    get width() {
        return MAZE_WIDTH;
    }

    get height() {
        return MAZE_HEIGHT;
    }
}

if (typeof module !== 'undefined' && module.exports) {
    module.exports = { Maze };
} else if (typeof window !== 'undefined') {
    window.Maze = Maze;
}
